File: island_sim/src/scoring.py
"""Local implementation of the competition scoring metric."""

from __future__ import annotations

import numpy as np

from .constants import ENTROPY_THRESHOLD, NUM_CLASSES


def entropy(probs: np.ndarray) -> np.ndarray:
    """Compute per-cell entropy. Input shape: (H, W, C). Output shape: (H, W)."""
    safe = np.clip(probs, 1e-15, 1.0)
    return -np.sum(probs * np.log(safe), axis=-1)


def kl_divergence(ground_truth: np.ndarray, prediction: np.ndarray) -> np.ndarray:
    """Compute per-cell KL(ground_truth || prediction). Shape: (H, W, C) → (H, W)."""
    gt = np.clip(ground_truth, 1e-15, 1.0)
    pred = np.clip(prediction, 1e-15, 1.0)
    return np.sum(gt * np.log(gt / pred), axis=-1)


def score_prediction(
    ground_truth: np.ndarray,
    prediction: np.ndarray,
) -> dict:
    """Score a prediction against ground truth using entropy-weighted KL divergence.

    Args:
        ground_truth: (H, W, 6) ground truth probability tensor.
        prediction: (H, W, 6) predicted probability tensor.

    Returns:
        Dict with score, weighted_kl, num_dynamic_cells, and per-cell details.
    """
    assert ground_truth.shape == prediction.shape
    assert ground_truth.shape[-1] == NUM_CLASSES

    cell_entropy = entropy(ground_truth)
    cell_kl = kl_divergence(ground_truth, prediction)

    # Dynamic cells are those with non-trivial entropy
    dynamic_mask = cell_entropy > ENTROPY_THRESHOLD
    num_dynamic = int(np.sum(dynamic_mask))

    if num_dynamic == 0:
        return {
            "score": 100.0,
            "weighted_kl": 0.0,
            "num_dynamic_cells": 0,
            "total_cells": ground_truth.shape[0] * ground_truth.shape[1],
        }

    # Entropy-weighted KL
    weighted_kl_sum = np.sum(cell_entropy[dynamic_mask] * cell_kl[dynamic_mask])
    entropy_sum = np.sum(cell_entropy[dynamic_mask])
    weighted_kl = weighted_kl_sum / entropy_sum

    # Score formula from docs
    score = max(0.0, min(100.0, 100.0 * np.exp(-3.0 * weighted_kl)))

    return {
        "score": score,
        "weighted_kl": float(weighted_kl),
        "num_dynamic_cells": num_dynamic,
        "total_cells": ground_truth.shape[0] * ground_truth.shape[1],
        "mean_entropy": float(np.mean(cell_entropy[dynamic_mask])),
        "mean_kl": float(np.mean(cell_kl[dynamic_mask])),
    }

File: island_sim/rust/src/scoring.rs
/// Scoring: entropy-weighted KL divergence (matches Python src/scoring.py).

use crate::types::NUM_CLASSES;

/// Score a prediction against ground truth.
/// Returns the competition score (0-100, higher is better).
pub fn score_prediction(ground_truth: &[f64], prediction: &[f64], height: usize, width: usize) -> f64 {
    // Must match Python src/constants.py ENTROPY_THRESHOLD
    let entropy_threshold = 0.01;
    let mut weighted_kl_sum = 0.0;
    let mut entropy_sum = 0.0;

    for y in 0..height {
        for x in 0..width {
            let base = (y * width + x) * NUM_CLASSES;

            // Compute entropy of ground truth cell
            let mut cell_entropy = 0.0;
            for c in 0..NUM_CLASSES {
                let p = ground_truth[base + c];
                if p > 1e-15 {
                    cell_entropy -= p * p.ln();
                }
            }

            if cell_entropy <= entropy_threshold {
                continue;
            }

            // Compute KL divergence for this cell
            let mut cell_kl = 0.0;
            for c in 0..NUM_CLASSES {
                let p = ground_truth[base + c].max(1e-15);
                let q = prediction[base + c].max(1e-15);
                cell_kl += p * (p / q).ln();
            }

            weighted_kl_sum += cell_entropy * cell_kl;
            entropy_sum += cell_entropy;
        }
    }

    if entropy_sum <= 0.0 {
        return 100.0;
    }

    let weighted_kl = weighted_kl_sum / entropy_sum;
    let score = 100.0 * (-3.0 * weighted_kl).exp();
    score.max(0.0).min(100.0)
}
